doc/guide.rst

```diff
@@ -201,6 +201,28 @@ What this does is tell pyOptSparse that the key ``obj_name`` in the function ret
 For optimizers that can do multi-objective optimization (e.g. NSGA2), multiple objectives can be added.
 Optimizers that can only handle one objective enforce that only a single objective is added to the optimization description.
 
+Analytic Derivatives
+++++++++++++++++++++
+
+pyOptSparse can automatically compute derivatives of the objective and constraint functions using finite differences or the complex-step method (See implementation details `here <https://mdolab-pyoptsparse.readthedocs-hosted.com/en/latest/api/gradient.html>`__).
+If analytic derivatives are available, users can define them within a user-defined function (let's use the function name ``sens()`` for the explanation purposes).
+This ``sens()`` function reads as inputs a dictionary containing design variables as well as another dictionary containing objective and constraint functions.
+Note that when supplying analytic derivatives, users need to make sure the keys in the ``sens()`` function are consistent with the keys of design variables (e.g. ``var_name``), constraints (e.g. ``equality_constraint``) and objectives (e.g. ``obj_name``).
+These names are simply examples used in previous paragraphs but they can be anything users prefer.
+Based on these keys (or names) users can separately define the analytic derivatives using a nested dictionary ``funcsSens`` (again, for the explanation convenience).
+Specifically, the first-layer keys should be associated with constraint and objective names while the second-layer keys are corresponding to design variables.
```

Comment:
It's important to mention here that you only need to specify non-zero sub-blocks. If `con1` does not depend on `x1` then you can leave that part blank, and pyOptSparse will assume it's zero.

Reply:
Oh, good to know. I just manually specify zero gradients. Will add this point.